### ISSUE TITLE
Different behavior between seemingly equivalent Py 2/3 files

### ISSUE THREAD

**Issue by bsolomon1124** [python/mypy#8600]
Version info:

```
$ python -V
Python 3.8.1
$ python -m mypy --version
mypy 0.770
```

Original file, `anystr.py`:

```python
#!/usr/bin/env python2
from typing import AnyStr

def concat(a, b):
    # type: (AnyStr, AnyStr) -> AnyStr
    return a + b

concat(u"foo", u"bar")  # Ok, output has type 'unicode'
concat(b"foo", b"bar")  # Ok, output has type 'bytes'
concat(u"foo", b"bar")  # Error, cannot mix unicode and bytes
```

The last line, [directly per the example in the typing docs](https://docs.python.org/3/library/typing.html#typing.AnyStr), should make mypy raise, but it does not with the `-2` flag:

```
$ python -m mypy -2 anystr.py 
Success: no issues found in 1 source file
$ python -m mypy --py2 anystr.py 
Success: no issues found in 1 source file
```

Yet, the Python3 version will:

```python
#!/usr/bin/env python3
from typing import AnyStr

def concat(a: AnyStr, b: AnyStr) -> AnyStr:
    return a + b

concat(u"foo", u"bar")  # Ok, output has type 'unicode'
concat(b"foo", b"bar")  # Ok, output has type 'bytes'
concat(u"foo", b"bar")  # Error, cannot mix unicode and bytes
```

```
$ python -m mypy anystr.py 
anystr.py:9: error: Value of type variable "AnyStr" of "concat" cannot be "object"
Found 1 error in 1 file (checked 1 source file)
```

Am I type-hinting or calling `mypy` incorrectly in the Python2 version?

**Comment by JelleZijlstra** [comment]
This is probably because mypy implicitly promotes str to unicode in Python 2.

**Comment by bsolomon1124** [comment]
Thanks @JelleZijlstra , I missed that note in the [mypy docs](https://mypy.readthedocs.io/en/stable/python2.html#type-checking-python-2-code) and seeing it now clears a lot up.
